core/commands/ptp.go

```diff
@@ -0,0 +1,392 @@
+		res.SetOutput(&PTPListenerInfoOutput{
+			Protocol: proto,
+			Address:  addr.String(),
+		})
+	},
+}
+
+var ptpDialCmd = &cmds.Command{
+	Helptext: cmds.HelpText{
+		Tagline: "Dial to a p2p listener.",
+
+		ShortDescription: `
+Establish a new connection to a peer service.
+
+When a connection is made to a peer service the ipfs daemon will setup one time
+TCP listener and return it's bind port, this way a dialing application can
+transparently connect to a p2p service.
+		`,
+	},
+	Arguments: []cmds.Argument{
+		cmds.StringArg("Peer", true, false, "Remote peer to connect to"),
+		cmds.StringArg("Protocol", true, false, "Protocol identifier."),
+		cmds.StringArg("BindAddress", false, false, "Address to listen for connection/s (default: /ip4/127.0.0.1/tcp/0)."),
+	},
+	Run: func(req cmds.Request, res cmds.Response) {
+		n, err := req.InvocContext().GetNode()
+		if err != nil {
+			res.SetError(err, cmds.ErrNormal)
+			return
+		}
+
+		err = checkEnabled(n)
+		if err != nil {
+			res.SetError(err, cmds.ErrNormal)
+			return
+		}
+
+		if !n.OnlineMode() {
+			res.SetError(errNotOnline, cmds.ErrClient)
+			return
+		}
+
+		addr, peer, err := ParsePeerParam(req.Arguments()[0])
+		if err != nil {
+			res.SetError(err, cmds.ErrNormal)
+			return
+		}
+
+		proto := "/ptp/" + req.Arguments()[1]
+
+		bindAddr, _ := ma.NewMultiaddr("/ip4/127.0.0.1/tcp/0")
+		if len(req.Arguments()) == 3 {
+			bindAddr, err = ma.NewMultiaddr(req.Arguments()[2])
+			if err != nil {
+				res.SetError(err, cmds.ErrNormal)
+				return
+			}
+		}
+
+		listenerInfo, err := n.PTP.Dial(n.Context(), addr, peer, proto, bindAddr)
+		if err != nil {
+			res.SetError(err, cmds.ErrNormal)
+			return
+		}
+
+		output := PTPListenerInfoOutput{
+			Protocol: listenerInfo.Protocol,
+			Address:  listenerInfo.Address.String(),
+		}
+
+		res.SetOutput(&output)
+	},
+}
+
+var ptpCloseCmd = &cmds.Command{
+	Helptext: cmds.HelpText{
+		Tagline: "Closes an active p2p stream or listener.",
+	},
+	Arguments: []cmds.Argument{
+		cmds.StringArg("Identifier", false, false, "Stream HandlerID or p2p listener protocol"),
+	},
+	Options: []cmds.Option{
+		cmds.BoolOption("all", "a", "Close all streams and listeners.").Default(false),
+	},
+	Run: func(req cmds.Request, res cmds.Response) {
+		n, err := req.InvocContext().GetNode()
+		if err != nil {
+			res.SetError(err, cmds.ErrNormal)
+			return
+		}
+
+		err = checkEnabled(n)
+		if err != nil {
+			res.SetError(err, cmds.ErrNormal)
+			return
+		}
+
+		if !n.OnlineMode() {
+			res.SetError(errNotOnline, cmds.ErrClient)
+			return
+		}
+
+		closeAll, _, _ := req.Option("all").Bool()
+
+		var proto string
+		var handlerID uint64
+
+		useHandlerID := false
+
+		if !closeAll {
+			if len(req.Arguments()) == 0 {
+				res.SetError(errors.New("no handlerID nor listener protocol specified"), cmds.ErrNormal)
+				return
+			}
+
+			handlerID, err = strconv.ParseUint(req.Arguments()[0], 10, 64)
+			if err != nil {
+				proto = "/ptp/" + req.Arguments()[0]
```

Comment:
What if someone creates protocol that is identified by decimal numbers?

If this is a problem right now, please fix it and create sharness test for it.